crates/circuit/src/gate_matrix.rs

```diff
@@ -0,0 +1,322 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use num_complex::Complex64;
+use std::f64::consts::FRAC_1_SQRT_2;
+
+pub static ONE_QUBIT_IDENTITY: [[Complex64; 2]; 2] = [
+    [Complex64::new(1., 0.), Complex64::new(0., 0.)],
+    [Complex64::new(0., 0.), Complex64::new(1., 0.)],
```

Comment:
it seems that in the matrices there are a few values that repeat themselves, maybe it will be cleaner to define them as consts?
e.g.  `ZERO` for `Complex64::new(0., 0.)`, `ONE` for `Complex64::new(1., 0.)`, then we can have `ONE_minus`, `I` and `I_minus` etc.

Reply:
I made a PR that does some things like this:
https://github.com/jlapeyre/qiskit-core/blob/new-some-gates-in-rust/crates/circuit/src/util.rs
https://github.com/jlapeyre/qiskit-core/blob/new-some-gates-in-rust/crates/circuit/src/gate_matrix.rs

This is part of a bigger PR. Of course adding these things could be organized differently. They could be moved to a smaller PR.

Reply:
In the most recent `num-complex` release they added built-in constants for `ZERO`, `ONE`, and `I`. When I original wrote these those didn't exist yet. But we use that version of num complex so it's simple enough to migrate to.

Reply:
Shelly: we can change these to `c64(1., 0.)` to make them a bit shorter (requires a new version of `num-complex`), which might help a bit.  It's totally possible to define a load of constants, but at some point, the words can be harder to read than just having the numbers in there; we've actually already had a bug in the Rust code because of us misreading a constant name for a variable.

Reply:
I noticed that some constants were present in newer versions of of num-complex. I tried but failed to use them, I dont recall the error. In any case, they are not exactly the constants that we need. I suppose we could add our own constants to them. But having one set of constants in one place looks cleaner to me.

 Also I wanted to use `c64` now and not wait till we use a newer version. I like the one I implemented because it allows `c64(1.5, 0)` while the built in one does not.  I also wrote a macro `c64!` because you can't use the functions when defining `static` and `const` variables.

I tried to make the constants/function/macro that I wrote consistent and compact. They are a total of 20 LOC. I think this is largely successful and that it makes the code more readable.

Reply:
I did this in https://github.com/Qiskit/qiskit/pull/12459/commits/faac6552778621a56285a20d422baaf94c3e819b as @jlapeyre points out `c64` is unfortunately not a `const fn` so we can't use it in statics. I just wrote a const fn so we could just do `c64(0., 1.)` and save us from needing `!`. A macro could potentially give us more flexibility so we could use different syntaxes and handle int literals. But I think this is ok, but it's simple enough to revert and save for a follow up if people would prefer that.